### PR TITLE
Java: Improve modelling of Spring requests, flow steps and XSS sinks

### PR DESCRIPTION
**Pull request by lcartey**
As part of some recent customer engagements I spent some time improving our modelling of Spring sources, flow steps and XSS sink.

The identification of trusted/untrusted data and XSS sinks is primarily taken from the [Spring reference documentation for request mapping handler methods](https://docs.spring.io/spring/docs/current/spring-framework-reference/web.html#mvc-ann-methods), which has a detailed breakdown of how parameters and return values are treated based on the type/annotations.

Review highlights:
 - This adds some general taint steps for flow out of `Map`s and `List`s. This is required because the tainted parameters to a Spring request method may be either of these types, and we need to track flow out of them. One concern here is whether this can result in spurious flow through these types, although I believe this will not happen in practice because we generally don't track flow into them (i.e. we don't have `List.add` etc. as a taint step).
 - I've also added a taint step for `String.replace`, as I had a customer provided benchmark which required flow through an inadequate `replace` call. If we retain this step, we may need to add some sanitizers to the XSS (and other) queries for "safe" looking replaces.
 - Spring has a mechanism whereby complex datatypes can automatically be populated from the request parameters or the request body (and thus contain user-provided data). For example, if parameters are annotated with `@RequestBody`:
    ```java
   @PostMapping
   public void createUser(@RequestBody User user) {
      ...
    }
    ```
    The `User` class will be populated from JSON provided in the request body, with fields populated recursively. To capture this aspect I have introduced the concept of `SpringUntrustedDataType`, and added taint flow steps through getter methods on these classes. I also wrote a helper (`stripType`) to unwrap generics (so that we can identify type parameters on lists etc. as also untrusted).

Work that still needs to be done:
 - [ ] Performance testing - I've only tried this on small to medium sized projects.
 - [ ] Verification that the additional taint steps do not create false positives in other projects.
 - [ ] Tidy up the QL

This addresses points 1-3 from: https://github.com/github/codeql-java-team/issues/9

### CODE REVIEW

**Comment by aschackmull** [comment]
> * This adds some general taint steps for flow out of `Map`s and `List`s. This is required because the tainted parameters to a Spring request method may be either of these types, and we need to track flow out of them. One concern here is whether this can result in spurious flow through these types, although I believe this will not happen in practice because we generally don't track flow into them (i.e. we don't have `List.add` etc. as a taint step).

This is not entirely accurate. We do already have default taint steps for flow in and out of collections and maps, see https://github.com/github/codeql/blob/master/java/ql/src/semmle/code/java/dataflow/internal/ContainerFlow.qll for details about what we currently include.

**Comment by aschackmull** [comment]
Differences job started: https://jenkins.internal.semmle.com/job/Changes/job/Java-Differences/775/

**Comment by aschackmull** [comment]
Could you rebase onto latest master, please? That would make it a bit easier to get a working Difference job running.

**Comment by aschackmull** [comment]
Annoyingly, rebases don't provide notifications.  In any case: https://jenkins.internal.semmle.com/job/Changes/job/Java-Differences/797/

**Comment by aschackmull** [comment]
The Differences job failed for some reason. Retrying: https://jenkins.internal.semmle.com/job/Changes/job/Java-Differences/815/

**Comment by aschackmull** [comment]
https://jenkins.internal.semmle.com/job/Changes/job/Java-Differences/822/

File: java/ql/src/semmle/code/java/dataflow/internal/TaintTrackingUtil.qll [hunk 1]
```diff
@@ -358,6 +376,36 @@ private predicate taintPreservingQualifierToMethod(Method m) {
...
   m = any(ProtobufMessageLite p).getAGetterMethod()
+  or
+  m instanceof MapMethod and
+  (
+    m.getName().regexpMatch("get|entrySet|keySet|values")
+  )
+  or
+  m.getDeclaringType().getSourceDeclaration().getASourceSupertype*().hasQualifiedName("java.util", "List") and
+  (
+    m.getName().regexpMatch("get|toArray|subList|spliterator|set|iterator|listIterator") or
```
Comment (aibaars) [inline java/ql/src/semmle/code/java/dataflow/internal/TaintTrackingUtil.qll:387]: @aschackmull Wouldn't it be better to replace these `regexMatch` calls with something like : 
```
m.hasName([ "get", "toArray", ...])
```

**Comment by aschackmull** [comment]
I've made a bunch of suggested changes here: https://github.com/lcartey/ql/pull/1

**Comment by aschackmull** [comment]
The differences job shows a 3% time increase for jdk, so that looks acceptable.

**Comment by aschackmull** [comment]
One of the individual commit comments (https://github.com/github/codeql/pull/3653/commits/7d555a7467d974de1c496f0868bc8e88ec994aff) suggests that some of the added steps are specific to XSS. Is this the case? If so, then we should move them to the XSS query.

**Comment by aschackmull** [comment]
3 tests are failing due to parameters marked with `@RequestParam` no longer is enough to be considered a taint source - the parameter must now also belong to a `SpringRequestMappingMethod`.  @lcartey could you verify that this is intentional? Because then I guess we'll just need to add `@RequestMapping` annotations in those 3 tests or something along those lines.

**Comment by lcartey** [comment]
> One of the individual commit comments ([7d555a7](https://github.com/github/codeql/commit/7d555a7467d974de1c496f0868bc8e88ec994aff)) suggests that some of the added steps are specific to XSS. Is this the case? If so, then we should move them to the XSS query.

I don't think they are specific to XSS, but they are most likely to be seen with XSS. I think they are fine as general taint steps.

**Comment by lcartey** [comment]
> 3 tests are failing due to parameters marked with `@RequestParam` no longer is enough to be considered a taint source - the parameter must now also belong to a `SpringRequestMappingMethod`. @lcartey could you verify that this is intentional? Because then I guess we'll just need to add `@RequestMapping` annotations in those 3 tests or something along those lines.

Yes, this is intentional. `@RequestParam` is not sufficient - there must be an `@RequestMapping` annotation (or similar) on the method itself, and a `@Controller` (or similar) on the declaring class. We have seen false positives for XSS caused by annotating dead `@RequestParam`s as untrusted data.

**Pull request merged**